src/components/multi_select_state_utils.js

```diff
@@ -40,12 +40,30 @@ export const getMinMaxIndexes = (currentIndex, firstItemShiftSelected) =>
 export const isWithin = (index, { minIndex, maxIndex }) =>
   index >= minIndex && index <= maxIndex;
 
-export const getNewSelectedItems = (itemId, items, selectedItems) => {
-  const sourceItems = items.filter(
-    item => item.id === itemId || findItem(item, selectedItems)
-  );
-  const destinationItems = selectedItems.filter(
+export const getNewSelectedItems = (
+  itemId,
+  items,
+  selectedItems,
+  keepSelectionOrder
+) => {
+  let alreadySelectedItems = [];
+  let sourceItems = [];
+  let destinationItems = [];
+
+  if (keepSelectionOrder) {
+    // In order to keep selection order on the list,
+    // First, iterate threw already selectedItems
+    alreadySelectedItems = selectedItems.filter(item => findItem(item, items));
```

Comment:
1. alreadySelectedItems is the same that contains in selectedItems, I think doesn't make seanse to do such filtering
2. your fix doesn't work when I use selection with Shift